Initial drop of nodereport project for nodejs repo

rnchamberlain opened the pull request:
Implementation of native NPM module for diagnostic report triggered on exception and fatal error events, signals and JavaScript API. Supports Node v4 and v6, Linux, Mac and Windows. Includes testcases and demo samples.

Fishrock123 commented:
I'm getting this on OS X 10.10.5 and Node.js v4/v6:

```
> node-gyp rebuild

  CXX(target) Release/obj.target/nodereport/src/node_report.o
../src/node_report.cc:675:76: warning: format specifies type 'long' but the argument has type '__darwin_suseconds_t' (aka 'int') [-Wformat]
    fprintf(fp,"\n  User mode CPU: %ld.%06ld secs", stats.ru_utime.tv_sec, stats.ru_utime.tv_usec);
                                       ~~~~~                               ^~~~~~~~~~~~~~~~~~~~~~
                                       %06d
../src/node_report.cc:676:78: warning: format specifies type 'long' but the argument has type '__darwin_suseconds_t' (aka 'int') [-Wformat]
    fprintf(fp,"\n  Kernel mode CPU: %ld.%06ld secs", stats.ru_stime.tv_sec, stats.ru_stime.tv_usec);
                                         ~~~~~                               ^~~~~~~~~~~~~~~~~~~~~~
                                         %06d
../src/node_report.cc:684:17: error: use of undeclared identifier 'RUSAGE_THREAD'
  if (getrusage(RUSAGE_THREAD, &stats) == 0) {
                ^
../src/node_report.cc:685:76: warning: format specifies type 'long' but the argument has type '__darwin_suseconds_t' (aka 'int') [-Wformat]
    fprintf(fp,"\n  User mode CPU: %ld.%06ld secs", stats.ru_utime.tv_sec, stats.ru_utime.tv_usec);
                                       ~~~~~                               ^~~~~~~~~~~~~~~~~~~~~~
                                       %06d
../src/node_report.cc:686:78: warning: format specifies type 'long' but the argument has type '__darwin_suseconds_t' (aka 'int') [-Wformat]
    fprintf(fp,"\n  Kernel mode CPU: %ld.%06ld secs", stats.ru_stime.tv_sec, stats.ru_stime.tv_usec);
                                         ~~~~~                               ^~~~~~~~~~~~~~~~~~~~~~
                                         %06d
4 warnings and 1 error generated.
make: *** [Release/obj.target/nodereport/src/node_report.o] Error 1
```

Also when applying this patch:
```
warning: squelched 20 whitespace errors
warning: 25 lines add whitespace errors.
```

mhdawson commented:
As an FYI Richard is on holiday this week , so response will be delayed.

mhdawson commented:
@bnoordhuis, believe you looked at PR to core, so you make be able to review this as the key issues will be the same, just as an npm versus part of core.

bnoordhuis commented on binding.gyp:
```diff
@@ -0,0 +1,30 @@
+      "conditions": [
+        ["OS=='linux'", {
+          "defines": [ "_GNU_SOURCE" ],
+          "cflags": [ "-g -O2 -std=c++11", ],
```
These should be split, not as a single string.

bnoordhuis commented on binding.gyp:
```diff
@@ -0,0 +1,30 @@
+    },
+  ],
+}
+
```
Inconsistent use of single and double quotes in this file.

sam-github replied:
^-- comment not addressed. js uses 4 SPC indents, c++ uses 2? I suggest pulling the eslint config from nodejs to keep some consistency of code, and make it easier for people to contribute consistent code to the project.

cjihrig replied:
+1 to using the ESLint config from core.

bnoordhuis commented on package.json:
```diff
@@ -0,0 +1,21 @@
+  "version": "0.1.0",
+  "keywords": [
+    ""
+  ],
```
Might as well leave out 'keywords' if you don't define any.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+ ******************************************************************************/
+NAN_METHOD(TriggerReport) {
+  Nan::HandleScope scope;
+  v8::Isolate* isolate = Isolate::GetCurrent();
```
You can use `info.GetIsolate()` instead of `Isolate::GetCurrent()`.  The former pulls it from the stack, the latter looks it up from the TLS.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+    } else {
+      Nan::ThrowSyntaxError("TriggerReport: filename parameter is too long (max 48 characters)");
+    }
+  }
```
Use a std::string so you don't have to impose a fairly arbitrary size limit?  At the very least compare against `sizeof(filename)` instead of a hard-coded value.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+    Nan::Callback callback(info[0].As<Function>());
+    // Creates a new Object on the V8 heap
+    Local<Object> obj = Object::New(isolate);
+    obj->Set(String::NewFromUtf8(isolate, "number"), Number::New(isolate, 54));
```
Why don't you use `Nan::New<...>` overloads for this?

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+  // If NodeReport newly requested for fatalerror, set up the V8 call-back
+  if ((nodereport_events & NR_FATALERROR) && !(previous_events & NR_FATALERROR)) {
+    isolate->SetFatalErrorHandler(OnFatalError);
+  }
```
Shouldn't you unset the handler if the flag is clear now but was set last time?

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+      uv_unref(reinterpret_cast<uv_handle_t*>(&nodereport_trigger_async));
+      RegisterSignalHandler(nodereport_signal, SignalDump, false);
+    }
+  }
```
You should check the return values of uv_sem_init() and uv_async_init() because those can fail (and sometimes will, i.e., it's not academic.)

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+ *
+ ******************************************************************************/
+NAN_METHOD(SetEvents) {
+  Nan::Utf8String parameter(info[0]->ToString());
```
Calling ToString() is not necessary when the value is fed to `Nan::Utf8String`.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+  if (nodereport_core) {
+    raise(SIGABRT); // core dump requested
+  } else {
+    exit(0); // no core dump requested
```
I don't think `exit(0)` is right for a fatal error.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+      if (nodereport_verbose) {
+        fprintf(stderr,"SignalDumpInterruptCallback - triggering NodeReport\n");
+      }
+      TriggerNodeReport(Isolate::GetCurrent(), kSignal_JS,
```
You get the isolate as a parameter, right?  No need to call `Isolate::GetCurrent()`.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+      // Request interrupt callback for running JavaScript code
+      isolate->RequestInterrupt(SignalDumpInterruptCallback, &report_signal);
+      // Event loop may be idle, so also request an async callback
+      size_t signo_data = static_cast<size_t>(report_signal);
```
I think passing (the address of) `report_signal` as a callback argument just obscures the fact that it's `report_signal`.  I'd drop it.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+  pthread_attr_t attr;
+  pthread_attr_init(&attr);
+
+  pthread_attr_setstacksize(&attr, PTHREAD_STACK_MIN);
```
This won't work on freebsd, PTHREAD_STACK_MIN is so small you can't run anything on it.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+  struct sigaction sa;
+  memset(&sa, 0, sizeof(sa));
+  sa.sa_handler = handler;
+  sa.sa_flags = reset_handler ? SA_RESETHAND : 0;
```
SA_RESETHAND doesn't work with some versions of freebsd's libthr.  The workaround is to set the handler to SIG_DFL in the signal handler.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+  }
+  const char* verbose_switch = getenv("NODEREPORT_VERBOSE");
+  if (verbose_switch != NULL) {
+	  nodereport_verbose = ProcessNodeReportVerboseSwitch(verbose_switch);
```
Indentation is off here.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+  const char* verbose_switch = getenv("NODEREPORT_VERBOSE");
+  if (verbose_switch != NULL) {
+	  nodereport_verbose = ProcessNodeReportVerboseSwitch(verbose_switch);
+  }
```
You should probably the getenv calls with a secure_getenv function that looks like this:
```c++
// Look up environment variable unless running as setuid root.
inline const char* secure_getenv(const char* key) {
#ifndef _WIN32
  if (getuid() != geteuid() || getgid() != getegid())
    return nullptr;
#endif
  return getenv(key);
}
```
(From node.cc.)

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+      uv_async_init(uv_default_loop(), &nodereport_trigger_async, SignalDumpAsyncCallback);
+      uv_unref(reinterpret_cast<uv_handle_t*>(&nodereport_trigger_async));
+      RegisterSignalHandler(nodereport_signal, SignalDump, false);
+    }
```
This block of code is duplicated here: https://github.com/nodejs/nodereport/pull/2/files#r77156266

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,338 @@
+  }
+}
+
+NODE_MODULE(nodereport, Initialize)
```
You should wrap the code in this file (and the header file) in a namespace.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,790 @@
+    return;
+  }
+  if (strlen(args) > NR_MAXPATH) {
+    fprintf(stderr, "Supplied nodereport directory path too long (max 2048 characters)\n");
```
Use `"max %zu characters", NR_MAXPATH` here?  (Or better: `sizeof(report_directory) - 1`.)

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,790 @@
+    fprintf(stderr, "Supplied nodereport directory path too long (max 2048 characters)\n");
+    return;
+  }
+  strcpy(report_directory, args);
```
It's not exactly wrong but `strcpy()` will set off most static analyzers.  I'd use `snprintf(report_directory, sizeof(report_directory), "%s", args)`.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,790 @@
+          tm_struct.tm_hour, tm_struct.tm_min, tm_struct.tm_sec);
+  fprintf(fp, "Module load time: %4d/%02d/%02d %02d:%02d:%02d\n",
+		  loadtime_tm_struct.tm_year+1900, loadtime_tm_struct.tm_mon+1, loadtime_tm_struct.tm_mday,
+		  loadtime_tm_struct.tm_hour, loadtime_tm_struct.tm_min, loadtime_tm_struct.tm_sec);
```
Indent is off here.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,790 @@
+  // Print Node.js and deps component versions
+  fprintf(fp, "\nNode.js version: %s\n", NODE_VERSION);
+  fprintf(fp, "(v8: %s, libuv: %s, zlib: %s, ares: %s)\n",
+        V8::GetVersion(), uv_version_string(), ZLIB_VERSION, ARES_VERSION_STR);
```
And here.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,790 @@
+    fprintf(fp,"Client\n");
+  }
+  TCHAR  infoBuf[256];
+  DWORD  bufCharCount = 256;
```
One space too many between the type and the variable name.  (Also, it's somewhat inconsistent to use camelCase here and snake_case everywhere else.)

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,790 @@
+  fprintf(fp, "\n==== Node.js libuv Handle Summary ==============================================\n");
+  fprintf(fp,"\n(Flags: R=Ref, A=Active, I=Internal)\n");
+  fprintf(fp,"\nFlags Type     Address\n");
+  uv_print_all_handles(NULL, fp);
```
Be aware that `uv_print_all_handles()` "...is meant for ad hoc debugging, there is no API/ABI stability guarantees."  (From the documentation.)

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,790 @@
+  if (static_cast<size_t>(info.vm_state) < arraysize(v8_states)) {
+      fprintf(fp, "JavaScript VM state: %s\n\n", v8_states[info.vm_state]);
+  } else {
+      fprintf(fp, "JavaScript VM state: <unknown>\n\n");
```
Indent is off here.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,790 @@
+  }
+  // Print the stack trace, adding in the pc values from GetStackSample() if available
+  for (int i = 0; i < stack->GetFrameCount(); i++) {
+    if ((size_t)i < info.frames_count) {
```
Prefer `static_cast<size_t>(i)`.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,790 @@
+  }
+
+  fprintf(fp, "\n\nEvent loop thread resource usage:");
+  if (getrusage(RUSAGE_THREAD, &stats) == 0) {
```
Only some platforms have RUSAGE_THREAD.  Try wrapping it in an `#ifdef RUSAGE_THREAD`.

I'm not 100% sure if it's guaranteed to be a define, POSIX only requires that it's a "symbolic constant", but on linux and freebsd it's a define (and those are the only two that have RUSAGE_THREAD, I think.  OS X doesn't have it but you can query it through proc_info(), I think.)

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,790 @@
+    env_var = *(environ + index++);
+  }
+
+static struct {
```
You could make this `const`.

bnoordhuis commented on src/node_report.h:
```diff
@@ -0,0 +1,38 @@
+using v8::Object;
+using v8::Number;
+using v8::String;
+using v8::Value;
```
You should really wrap this in a namespace.  `using` directives in the global namespace in header files are a perennial source of name conflicts.

rnchamberlain commented:
@Fishrock123 the two commits above should fix the compilation warnings + error you saw on OS X. I'll also fix the whitespace errors shortly, thanks.

Fishrock123 commented:
Getting this now: (OS X 10.10.5)

```
> nodereport@0.1.0 install /Users/Jeremiah/Documents/nodesource/nodereport
> node-gyp rebuild

  CXX(target) Release/obj.target/nodereport/src/node_report.o
  CXX(target) Release/obj.target/nodereport/src/module.o
../src/module.cc:288:32: error: use of undeclared identifier 'secure_getenv'
  const char* verbose_switch = secure_getenv("NODEREPORT_VERBOSE");
                               ^
../src/module.cc:292:32: error: use of undeclared identifier 'secure_getenv'
  const char* trigger_events = secure_getenv("NODEREPORT_EVENTS");
                               ^
../src/module.cc:296:34: error: use of undeclared identifier 'secure_getenv'
  const char* core_dump_switch = secure_getenv("NODEREPORT_COREDUMP");
                                 ^
../src/module.cc:300:32: error: use of undeclared identifier 'secure_getenv'
  const char* trigger_signal = secure_getenv("NODEREPORT_SIGNAL");
                               ^
../src/module.cc:304:29: error: use of undeclared identifier 'secure_getenv'
  const char* report_name = secure_getenv("NODEREPORT_FILENAME");
                            ^
../src/module.cc:308:32: error: use of undeclared identifier 'secure_getenv'
  const char* directory_name = secure_getenv("NODEREPORT_DIRECTORY");
                               ^
6 errors generated.
make: *** [Release/obj.target/nodereport/src/module.o] Error 1
```

rnchamberlain commented:
@Fishrock123 darn, I really need a Mac to hand, plse try fix above

mhdawson commented:
@rnchamberlain you might try logging into our internal IBM OSX machines to do test builds there.

bnoordhuis commented:
@rnchamberlain You can just copy the secure_getenv implementation from node.cc, it's a glibc-ism otherwise.  Granted, the glibc version also checks capabilities, node does not.

rnchamberlain commented:
@bnoordhuis I found that nan.h pulled in the glibc secure_getenv implementation via stdlib.h and that gave a clash with a local one copied from node.cc. So I went for the official one on Linux, as it has a bit more functionality. But maybe some Linux installations would not have that level of glibc. I guess it is worth having the local implementation for OSX anyway. I'll have a think.

bnoordhuis commented:
> I found that nan.h pulled in the glibc secure_getenv implementation via stdlib.h and that gave a clash with a local one copied from node.cc.

And that is why you should use namespaces. :-)

rnchamberlain commented:
@bnoordhuis thanks for the comprehensive code review, I think I have covered the points now. 

A couple of notes:

1. Un-setting isolate->SetFatalErrorHandler(). There does not seem to be a way to 'unset' this
callback to revert to the default v8 behaviour. I tried re-setting it with a NULL pointer for the
callback, but the v8 behaviour is then a bit different (and less satisfactory) than the default, see code
in Utils::ReportApiFailure().

2. secure_getenv(). I'd rather avoid having a local implementation, so have gone with using the
glibc implementation if it is available, else simple getenv()

joshgav commented:
Just started reviewing :) <br><br>:octocat: *Sent from [GH](http://nodegh.io).*

mhdawson commented:
Is there a way that we can also provide command line options? Not necessary for the first pass but could be useful.

mhdawson commented on demo/exception.js:
```diff
@@ -0,0 +1,34 @@
+console.log('exception.js: Node running');
+console.log('exception.js: Go to http://<machine>:8080/ or http://localhost:8080/');
+
+setTimeout(function(){
```
Missing space before {

mhdawson commented on demo/loop.js:
```diff
@@ -0,0 +1,55 @@
+console.log('loop.js: Node running');
+console.log('loop.js: Go to http://<machine>:8080/ or http://localhost:8080/');
+
+setTimeout(function(){
```
missing space before {

mhdawson commented on src/node_report.cc:
```diff
@@ -0,0 +1,785 @@
+    break;
+  }  // end switch(event)
+#endif
+  fflush(fp);
```
Should the code above be a separate function like the ones below.

mhdawson commented on package.json:
```diff
@@ -0,0 +1,18 @@
+    "nan": "^2.3.5"
+  },
+  "license": "MIT"
+}
```
We should add the scripts entry so that the tests are run when you type 
npm test

mhdawson commented:
LGTM with a few nits.

rnchamberlain commented:
@mhdawson thanks for the review, I have moved the blocks of code that print version information and stack traces in node_report.cc into separate functions, and fixed the minor nits. Also added the scripts entry into package.json so uses can run "npm test", and fixed some CRLF line endings in the test code.

I had not considered line-mode options (options are currently supported via env vars and/or JS API calls). A typical node command line is:
```
node <node and v8 options> application.js <application options>
```
We could allow nodereport options to be mixed in with either the node/v8 or the application options. Suitable unique naming would be need, e.g something like "nodereport_events=exception+signal". Application options are available via process.argv, though I am not sure how to access that from C++, otherwise I should be able to get the command line from the OS . Anyway, something for a future enhancement.

Fishrock123 commented:
Got it to build on OS X, is `NodeReport.20161004.120508.2622.001.txt` just meant to be read manually? Seems reasonably understandable to me.

Fishrock123 commented:
Should this be merged so that it can be worked on more collaboratively?

rnchamberlain commented:
@Fishrock123 yes absolutely, the report is intended to be read manually, using any text editor/notepad/email etc, and should be readily understandable to node developers, admins, testers etc

Yes this is ready to go and for others to contribute. @mhdawson - not sure how this repo is set up, who has powers to merge this? thanks

Fishrock123 commented:
I think we should probably just give you perms to merge it. cc/ @nodejs/post-mortem

misterdjules commented:
This repo has @mhdawson and the @nodejs/post-mortem team as collaborators. @rnchamberlain isn't (to my surprise) a member of the @nodejs/post-mortem team, even though he's clearly active in that space, so I added him to that team.

Once @rnchamberlain accepts the invite, he'll have the credentials to push commits to this repo.

sam-github commented on README.md:
```diff
@@ -8,3 +8,61 @@ It includes Javascript and native stack traces, heap statistics,
+
+    var nodereport = require('nodereport');
+
+By default, this will trigger a NodeReport to be written to the current
```
I don't think this is a good idea. It means that there is no way to opt-out of auto-handling of `unhandledException`, which would be useful for someone who wanted to only use the manual `.triggerReport()`. I think it would make more sense for a require of `nodereport` to make no changes to global process state until it was attached. Syntax could be `require('nodereport')()`, or `require('nodereport').attach()`, or any other variant.

sam-github replied:
also, I would want to control independently whether nodereport attached to SIGUSR2, or even configure the signal.

rnchamberlain replied:
Hi @sam-github, you can override the defaults and configure which events you want via the API, so
`require("nodereport").setEvents("fatalerror+signal");`
will switch off the exception handler, likewise
`require("nodereport").setEvents("exception+fatalerror");`
will switch off the signal handler.

`require("nodereport").setSignal("SIGQUIT");`
will configure SIGQUIT for the trigger signal instead of SIGUSR2, and we could add more choices for that (though I just found a bug in that code, fix on it's way).

joshgav replied:
Perhaps an optional options arg with key `events` as an array of potential triggers and key `sigusr2` as a boolean to enable/disable the signal would be a nice approach.

Such an arg would be reasonably extensible for other triggers too.

Also allows for clear defaults to be set early in code.

sam-github commented on README.md:
```diff
@@ -8,3 +8,61 @@ It includes Javascript and native stack traces, heap statistics,
+Configuration is available via these environment variables and/or Javascript
+API calls:
+
+    export NODEREPORT_EVENTS=exception+fatalerror+signal
```
that its possible to configure nodereport via env vars is great, that its necessary to do so is I think too much focus on policy instead of mechanism

rnchamberlain replied:
I'd expect the `.setEvents()` API to be the usual way to configure it, with the env vars as an alternative maybe for admins when changing the code is not convenient. I'll update the readme to reflect that. 

Re the defaults, the design was for all the events to work 'out of the box' i.e. with just require("nodereport"). We could take a more cautious approach as you suggest, and just have 
`.triggerReport()` available until  `.setEvents(list of selected events)` is called.

Fishrock123 commented:
I don't think we should pull from core eslint. Just use `standard`.

mhdawson commented:
I believe I have given everybody in the post-mortem team write access to this repo. Was just waiting for 1 more L.G.T.M from one of the other reviewers before landing.

joshgav commented on src/module.cc:
```diff
@@ -0,0 +1,369 @@
+}
+NAN_METHOD(SetCoreDump) {
+  Nan::Utf8String parameter(info[0]);
+  nodereport_events = ProcessNodeReportCoreSwitch(*parameter);
```
I believe this is intended: `s/nodereport_events/nodereport_core`

rnchamberlain replied:
@joshgav good spot, and same error in SetSignal method below, fix in progress

Fishrock123 commented:
Status? I think I'd actually like to start using this.

rnchamberlain commented:
Fixed as discussed with @sam-github and @joshgav. So `require('nodereport')` by default now just enables the `.triggerReport()` functionality. No changes to the Node/v8 hooks and signals until the requirer explicitly calls  `.setEvents("exception+signal+apicall+etc")`.

Status: as per @mhdawson above, just another LGTM then we land. Meantime available to try out now with
 `npm install rnchamberlain/nodereport-1`

sam-github commented:
@bnoordhuis LGTY?

@rnchamberlain API LGTM. If you want to make it easy to enable all capabilities, you could implement a usage like:

```
require('nodereport')();
```

So a side-effectless require is possible, but all side-effects can be opted in easily, and the "module call" above would check for env vars

You could also extend above with an options object rather than method calls for configuration:

```
require('nodereport')({
  events: 'signal',
  signal: 'SIGHUP',
  });
```

rnchamberlain commented:
@sam-github I had a look at both of your suggested syntax options, but couldn't immediately work out  how to code them on the C++ side! There is a mention of a two-argument form of the Init() function here: https://nodejs.org/api/addons.html, which has a module object passed in as well as the exports object? Anyway yes, potential future enhancement - thanks.

bnoordhuis commented on .gitignore:
```diff
@@ -0,0 +1,38 @@
+
+# Build directories
+/build/
+/node_modules/
```
This file uses Windows line endings for some reason.

bnoordhuis commented on README.md:
```diff
@@ -1,10 +1,66 @@
+    api.js - NodeReport triggered by Javascript API call
+    exception.js - NodeReport triggered by unhandled exception
+    fatalerror.js - NodeReport triggered by fatal error on Javascript heap out of memory
+    loop.js - looping application, NodeReport triggered using kill -USR2 <pid>
```
There is some trailing whitespace in this file.

bnoordhuis commented on demo/loop.js:
```diff
@@ -0,0 +1,53 @@
+setTimeout(function() {
+  console.log('loop.js: timeout expired, exiting.');
+  process.exit(0); 
+}, 60000);
```
Ditto.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,388 @@
+
+  if (info[0]->IsString()) {
+    // Filename parameter supplied
+    Nan::Utf8String filename_parameter(info[0]->ToString());
```
You don't have to ToString() the argument to Nan::Utf8String.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,388 @@
+    if (filename_parameter.length() < NR_MAXNAME) {
+      snprintf(filename, sizeof(filename), "%s", *filename_parameter);
+    } else {
+      Nan::ThrowSyntaxError("nodereport: filename parameter is too long");
```
SyntaxError is kind of weird to throw for a bad argument.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,388 @@
+ ******************************************************************************/
+NAN_METHOD(SetEvents) {
+  Nan::Utf8String parameter(info[0]);
+  v8::Isolate* isolate = node_isolate;
```
Why don't you use info.GetIsolate() here?

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,388 @@
+bool OnUncaughtException(v8::Isolate* isolate) {
+   // Trigger NodeReport if required
+  if (nodereport_events & NR_EXCEPTION) {
+    TriggerNodeReport(isolate, kException, "exception", "OnUncaughtException (nodereport/src/module.cc)", NULL);
```
`__FILE__`?

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,388 @@
+  } else {
+    return false;
+  }
+}
```
Just `return nodereport_core;`?

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,388 @@
+}
+
+#ifndef _WIN32
+static void SignalDumpInterruptCallback(Isolate *isolate, void *data) {
```
Somewhat schizophrenic star style.  It leans to the right here but to the left in most other places.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,388 @@
+static void SignalDumpAsyncCallback(uv_async_t* handle) {
+  if (report_signal != 0) {
+    if (nodereport_verbose) {
+      fprintf(stdout,"nodereport: SignalDumpAsyncCallback handling signal\n");
```
Space after comma.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,388 @@
+    }
+    if (nodereport_events & NR_SIGNAL) {
+      if (nodereport_verbose) {
+        fprintf(stdout,"nodereport: SignalDumpAsyncCallback triggering NodeReport\n");
```
Ditto.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,388 @@
+// Utility function to restore an OS signal handler to its previous setting
+static void RestoreSignalHandler(int signal) {
+  sigaction(signal, &saved_sa, nullptr);
+}
```
It's arguably better API design to have the caller pass in `saved_sa` explicitly in RegisterSignalHandler() and RestoreSignalHandler().  It's easy to get fooled into thinking that it restores the handler for signal `signal` but that's not exactly what it does.

bnoordhuis replied:
As an aside: maybe `s/signal/signo/` because that a) doesn't shadow the libc function of the same name, and b) is consistent with the function below.

bnoordhuis commented on src/module.cc:
```diff
@@ -0,0 +1,388 @@
+NODE_MODULE(nodereport, Initialize)
+
+}  // namespace nodereport
+
```
General observation: there is a somewhat inconsistent mix of nullptr vs. NULL in this file.  Since you're going for C++11, I'd use nullptr everywhere.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,808 @@
+#endif
+static void PrintGCStatistics(FILE *fp, Isolate* isolate);
+static void PrintSystemInformation(FILE *fp, Isolate* isolate);
+static void WriteInteger(FILE *fp, size_t value);
```
The star-leans-left/star-leans-right dichotomy exists in this file as well.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,808 @@
+    }
+  }
+  return event_flags;
+}
```
Indentation seems off in this function.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,808 @@
+             ".%4d%02d%02d.%02d%02d%02d.%d.%03d.txt",
+             tm_struct.wYear, tm_struct.wMonth, tm_struct.wDay,
+             tm_struct.wHour, tm_struct.wMinute, tm_struct.wSecond,
+            pid, seq);
```
Indentation.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,808 @@
+      strcat(pathname, "\\");
+#else
+      strcat(pathname, "/");
+#endif
```
Potential buffer overrun here.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,808 @@
+
+  fprintf(stderr, "Node.js report completed\n");
+  if (name != NULL) {
+    snprintf(name, NR_MAXNAME + 1, "%s", filename);  // return the NodeReport file name
```
Declare `name` like this in the function's parameter list: `char (*name)[NR_MAXNAME + 1]`.  That way the caller cannot pass in a wrong-sized buffer.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,808 @@
+  fprintf(fp,"\nOS version: Windows ");
+#if defined(_MSC_VER) && (_MSC_VER >= 1900)
+  if (IsWindows1OrGreater()) {
+    fprintf(fp,"10 ");
```
Space after comma, here and elsewhere.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,808 @@
+    break;
+  case kFatalError:
+#if NODE_VERSION_AT_LEAST(6, 0, 0)
+    if (!strncmp(location,"MarkCompactCollector", sizeof("MarkCompactCollector") - 1)) {
```
Ditto.

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,808 @@
+  // Walk the frames printing symbolic information if available
+  for (int i = 0; i < numberOfFrames; i++) {
+    DWORD64 dwOffset64 = 0;
+    DWORD64 dwAddress = (DWORD64)(frames[i]);
```
C++-style casts, please!

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,808 @@
+    if (SymFromAddr(hProcess, dwAddress, &dwOffset64, pSymbol)) {
+        DWORD dwOffset = 0;
+        IMAGEHLP_LINE64 line;
+        line.SizeOfStruct = sizeof(IMAGEHLP_LINE64);
```
`sizeof(line)`?

rnchamberlain replied:
https://msdn.microsoft.com/en-us/library/windows/desktop/ms680190%28v=vs.85%29.aspx

bnoordhuis replied:
Right, but that doesn't preclude `sizeof(line)`, does it?

rnchamberlain replied:
Either sizeof(<type>) or sizeof(<variable>) work OK I believe, the MSDN examples uses sizeof(<type>). Is sizeof(<variable>) preferred?

bnoordhuis commented on src/node_report.cc:
```diff
@@ -0,0 +1,808 @@
+  lpvEnv = GetEnvironmentStrings();
+  if (lpvEnv != NULL) {
+    // Variable strings are separated by NULL byte, and the block is terminated by a NULL byte.
+    lpszVariable = (LPTSTR) lpvEnv;
```
Ditto.

bnoordhuis commented on src/node_report.h:
```diff
@@ -0,0 +1,75 @@
+#endif  // !__GLIBC_PREREQ(2, 17)
+#else
+#define secure_getenv getenv
+#endif  // defined(__GLIBC__)
```
Potential downside is that a .node compiled against glibc 2.17 won't work with e.g. glibc 2.15.  That is, you won't be able to simply copy the shared object around.

rnchamberlain replied:
I wonder if that's a common thing to do, copy binaries around from an npm install. Someone is bound to do it I suppose, and they will get an unresolved symbol at load time. Under consideration! 

I think I have fixed everything else, see commits below, and thanks for the review.

sam-github replied:
Its a nice to have to be able to build on one system, and run on another slightly different one. It fails often enough that its usually necessary to build on a system of identical os/etc level and architecture as the deployment/production system.

rnchamberlain commented:
Looking back we have an LGTM and an API LGTM on this, I'll merge shortly unless anyone objects. Thanks.
